File: server client communication/client_mqtt.js
// Environment setup
// npm install mqtt

// Code run
// node client.js

// Run `npm install mqtt` to install the MQTT library
const mqtt = require('mqtt');

// Broker information
const brokerAddress = 'mqtt://broker.hivemq.com';
const receiveTopic = 'test/cs489/1';
const sendTopic = 'test/cs489/2';

// Connect to the broker
const client = mqtt.connect(brokerAddress);

// Event handler when the client is connected to the broker
client.on('connect', () => {
    console.log('Client connected to broker');

    // Subscribe to the topic to receive messages
    client.subscribe(receiveTopic, (err) => {
        if (!err) {
            console.log(`Subscribed to topic: ${receiveTopic}`);
        } else {
            console.error('Subscription error:', err);
        }
    });
});

// Event handler when the client receives a message
client.on('message', (topic, message) => {
    console.log(`\nClient received: ${message.toString()}`);
});

// Function to send messages from the command line
const readline = require('readline').createInterface({
    input: process.stdin,
    output: process.stdout
});

const promptInput = () => {
    readline.question('Enter a message to send to the server: ', (input) => {
        if (input.toLowerCase() === 'exit') {
            console.log('Exiting...');
            client.end(); // Disconnect the client
            readline.close(); // Close the input interface
            return;
        }
        client.publish(sendTopic, input);
        console.log(`Client sent: ${input}`);
        promptInput(); // Recursively prompt for the next input
    });
};

// Start the prompt loop
promptInput();

File: server client communication/client_mqtt.py
# pip install paho-mqtt==1.6.1
import paho.mqtt.client as mqtt

# Broker information
broker_address = "broker.hivemq.com"
port = 1883
receive_topic = "test/cs489/1"
send_topic = "test/cs489/2"

# Callback when the client receives a message
def on_message(client, userdata, message):
    received_message = message.payload.decode('utf-8')
    print(f"\nClient received: {received_message}")

# Create MQTT client instance and set callback
client = mqtt.Client("Client")
client.on_message = on_message

# Connect to the broker
client.connect(broker_address, port)

# Subscribe to the topic where echo messages are received
client.subscribe(receive_topic)

# Start the loop in a separate thread to process network traffic and dispatch callbacks
client.loop_start()

# Continuously get input from CLI and send it to the server
try:
    while True:
        message = input("Enter a message to send to the server: ")
        if message.lower() == 'exit':
            break  # Exit the loop if 'exit' is typed
        client.publish(send_topic, message)
        print(f"Client sent: {message}")
except KeyboardInterrupt:
    print("\nExiting...")

# Stop the loop and disconnect cleanly
client.loop_stop()
client.disconnect()
